### ISSUE TITLE
SAF values aren't always consistent with actual reads?

### ISSUE THREAD

**Issue by TeresaPegan** [ANGSD/angsd#445]
Hello! I'm trying to understand how -dosaf works, and I have been comparing the dosaf values with the actual reads and base qualities at a given locus. I have made a SAF file using a single diploid individual. (I am calculating Fst between two diploid individuals).

I am noticing some things that do not make sense to me. For example, here is the SAF output for some example loci:

```
CM027507.1	1853613	       0.000000.       -0.693156	-6.907755
CM027507.1	3952991	        -7.881721	-0.692904	0.000000
CM027507.1	19891662	-6.956041	0.000000	-8.054336

```

This is a single diploid individual, so there are only 3 possible allele frequencies and these correspond to the genotypes of the individual. Since 0 is the largest value for each, it seems like the first locus is most likely homozygous for one allele, the second locus most likely homozygous for the other allele, and the third is most likely a heterozygote.

It's also worth noting that the first two loci also have fairly high likelihoods for being heterozygous (the likelihood is relatively close to 0). It seems like ANGSD is a bit ambivalent about whether the first two loci are heterozygous or homozygous.

Now, here are the actual reads and quality scores for these three loci, from samtools mpileup (without the reference):

```
CM027507.1	1853613	        N	2	aa	JJ
CM027507.1	3952991	        N	2	Ag	JJ
CM027507.1	19891662	N	2	Ct	JJ

```

All of these loci have bases with the exact same coverage and quality.

**I can't understand why -dosaf is treating locus 3952991 as homozygous** (i.e., that the allele frequency is highest in the third column). It assigns the same likelihood (-0.69) for the heterozygous condition to locus 3952991, which shows an "a" and a "g", as to locus 1853613, which only has reads showing an "a". Locus 19891662 shows two different base types, just like 3952991 -- and it has exactly the same quality scores -- yet in this case the heterozygous condition clearly has the highest likelihood.

Also, here are the genotype likelihoods for this individual from a beagle file also generated by ANGSD. These correspond to what I expect based on the reads.

```
CM027507.1	1853613   0.799992 0.200008 0.000000
CM027507.1	3952991	  0.000106 0.999788 0.000106
CM027507.1	19891662   0.000106 0.999788 0.000106
```
And also, the mpileup output in relation to my reference genome, which demonstrates that each of these sites has at least one base matching the reference base:

```
CM027507.1	1853613	        A	2	,,	JJ
CM027507.1	3952991	        A	2	.g	JJ
CM027507.1	19891662	t	2	C,	JJ
```

Does anyone have insights into how -dosaf generates its values that might help explain this? Thanks!

Here is the code I used to generate the SAF:

```
$ANGSDIR/angsd -b flinflon/$FOLD/ANGSD_FST/$BASE"_path" -ref $ANC -anc $ANC -out $OUTDIR/$SP"_"$POP"_"$SELECT$BASE \
-dosaf 1 -GL 1 -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -minMapQ 20 -minQ 20 \
-doCounts 1 -C 50 -sites $SITES 
```

Version of ANGSD: 0.935
(I have to use older versions because I get a lot of segfaults in more up-to-date versions of ANGSD).

**Comment by TeresaPegan** [comment]
Just a quick note to say that ANGSD 0.937 is still showing the behavior described above. I updated ANGSD, reran all of my analyses, and I still see the exact same thing going on as described above. The individual has read A and G at locus CM027507.1:3952991 and yet in the saf file ANGSD seems to consider the highest likelihood allele frequency at this locus to be 2 (homozygous minor), not 1 (heterozygous).
```
CM027507.1	3952991	-7.881721	-0.692904	0.000000
```


 **However**, I get an expected result when I use -dosaf 4 and use the beagle file for my individual of interest as input. 
```
CM027507.1	3952991	-0.999682	0.000000	-0.999682
```

I'll be using -dosaf 4 from now on. Here is the code that worked. I did not need to use my $SITES argument because the beagle file was already filtered to the sites I wanted.

```
$ANGSDIR/angsd -beagle $INBEAG -ref $ANC -anc $ANC -fai $FAI -out $OUTDIR/Sfusc_UP17035_testsaf \
-dosaf 4 -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -minMapQ 20 -minQ 20 -C 50  
```

I guess I could close this because using -dosaf 4 solves my problem, but it does seem like there is some kind of problem with -dosaf 1 so I'll leave it open for now.

**Comment by TeresaPegan** [comment]
I did just notice that I had used -GL 2 for creating my beagle file (which I used with -dosaf 4), versus -GL 1 when I was making the SAF using -dosaf 1. Perhaps this accounts for the difference -- although I still don't really understand why any GL method would produce the results I got.

**Comment by ANGSD** [comment]
Hi Teresa, Could you try do add -doglf 4 and redo the analysis just for these sites. Like
echo -en "CM027507.1\t1853613\nCM027507.1\t3952991\nCM027507.1\t19891662\n" >rf.txt

$ANGSDIR/angsd -b flinflon/$FOLD/ANGSD_FST/$BASE"_path" -ref $ANC -anc $ANC -out $OUTDIR/$SP"_"$POP"_"$SELECT$BASE \
-dosaf 1 -GL 1 -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -minMapQ 20 -minQ 20 \
-doCounts 1 -C 50 -rf rf.txt

This should emit all 10gls for the individual and we should be able to see exactly what is happening.

**Comment by TeresaPegan** [comment]
Hi,
Thanks for the tip! I tried this, creating the rf file and then running the code you pasted (with -dosaf 1 and my bam file as input). However, this doesn't give me the expected results. For one thing, it prints out the entire genome in the SAF, not just the three sites in rf.txt. Also, the resulting saf doesn't have 10 gls, just the three possible allele frequencies again (and I get the same result as before for locus 3952991). Are you suggesting that I try to make a beagle file with -doGlf 1 to get all 10 gls?
Sorry for the confusion...
Thanks,
-Teresa

```
echo -en "CM027507.1\t1853613\nCM027507.1\t3952991\nCM027507.1\t19891662\n" >rf.txt

$ANGSDIR/angsd -b flinflon/$FOLD/ANGSD_FST/$BASE"_path" -ref $ANC -anc $ANC -out $OUTDIR/Sfusc_UP17035_testsaf10gls \
-dosaf 1 -GL 1 -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -minMapQ 20 -minQ 20 \
-doCounts 1 -C 50 -rf rf.txt

```

**Comment by TeresaPegan** [comment]
I just confirmed that -dosaf 1 will produce the same unexpected result with -GL 2, so the GL method is not the problem:

```
$ANGSDIR/angsd -b flinflon/$FOLD/ANGSD_FST/$BASE"_path" -ref $ANC -anc $ANC -out $OUTDIR/Sfusc_UP17035_testsafGL2 -dosaf 1 -GL 2 -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -minMapQ 20 -minQ 20 -doCounts 1 -C 50 -sites $SITES
```

```
CM027507.1	3952991	-9.440573	-0.693068	0.000000
```

I have now also made two files emitting all ten genotypes (not SAF files, GL files). Here is the result from the one with -GL 2:
```
CM027507.1      3952991 -9.152890       -9.846011       0.000000        -9.846011       -19.692022      -9.846011       -19.692022      -9.152890       -9.846011       -19.692022
```

Here is the one with -GL 1:
```
CM027507.1      3952991 -8.054335       -8.747491       0.000000        -8.747491       -16.801827      -8.747491       -16.801827      -8.054335       -8.747491       -16.801827
```

These results are pretty much the same. It looks like they produce the expected result, where the third genotype (alphabetically, which the manual says is the order of the 10 gls) is AG, which is the genotype I expect to have the highest likelihood.

So, the problem really does seem to be associated with -dosaf. I am interpreting the SAF correctly, right? When the SAF has the 0 in its 3rd position, that indicates it thinks the minor allele frequency is 2? (For my diploid, the first SAF position corresponds to minor allele frequency = 0, and the second position to minor allele frequency = 1)? In any case, I expect the highest likelihood in the SAF to be in the second position, for allele frequency = 1, because the reads here are an A and a G.

Thanks!
-Teresa

**Comment by ANGSD** [comment]
Hello Teresa sorry for wrong help.
When you generate the input for the -rf it should not be formattted with tab but with ":"
So you should do this instead:
echo -en "CM027507.1:1853613\nCM027507.1:3952991\nCM027507.1:19891662\n" >rf.txt

Yes, the order of the gls are
AA,AC,AG,AT,CC,CG,CT,GG,GT,TT

I of course interested in understanding what is happening here, and I was thinking it might be faster and easier if you supply me with the reads that overlap this specific site and then I can look into it.

something like:

samtools view your.bam CM027507.1:3952991-3952991 -b >subs.bam

**Comment by TeresaPegan** [comment]
Ah, ok. Of course, here are the two reads. Thanks so much for looking into it!

I am pasting the two reads here as text because github wouldn't let me attach them in bam format. Is that ok?

Thanks,
-Teresa

```
K00124:646:HFHVVBBXX:6:1104:14225:69098	99	CM027507.1	3952937	60	4S102M44S	=	3953039	252	GACAGATCCTACCCCTTGCTCAAATTGTAGCAGCATTTCAGTAATGGGAGTTTCCTTCATGGGAGTGAAGTTCCTCTGAGTTGCCAGCACTTTTCACTCATGCAGCTTCTTAGGTTTAAGTGACAACACTAAAGAGAAAACTCCTGGGTG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFFJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJ	MC:Z:150M	MD:Z:29A116	PG:Z:MarkDuplicates	RG:Z:1	NM:i:1	MQ:i:60	AS:i:141	XS:i:21
K00124:646:HFHVVBBXX:6:2205:5193:59675	83	CM027507.1	3952982	60	128M2D22M	=	3952938	-196	GTTTCCTTCGTGGGAGTGAAGTTCCTCTGAGTTGCCAGCACTTTTCACTCATGCAGCTTCTTAGGTTTAAGTGACAACACTAAAGAGAAAACTCCTGGGTGTCTCTATAAAATAGCTTGATAAACGTAATTTGGTTTATGAGGATTGAAC	JJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFFFJFJJJJJJJJJJJJF<AAA	MC:Z:44M106S	MD:Z:9A115A2^TG22PG:Z:MarkDuplicates	RG:Z:1	NM:i:4	MQ:i:60	AS:i:132	XS:i:19
```

**Comment by nspope** [comment]
@TeresaPegan could you please email me the bam file (with header) containing the overlapping reads, and the fasta for the contig that they are mapped to? nspope at utexas dot edu

**Comment by Sol2Man** [comment]
Was this issue resolved?

**Comment by nspope** [comment]
Hmm, I can't replicate the problem with github head of ANGSD. Here is example using the values provided by @TeresaPegan above for -GL 2:
```bash
##from https://github.com/ANGSD/angsd/issues/445#issuecomment-1049127693
#CM027507.1      3952991 -9.152890       -9.846011       0.000000        -9.846011       -19.692022      -9.846011       -19.692022      -9.152890       -9.846011       -19.692022

echo '
out <- c(-9.152890,-9.846011,0.000000,-9.846011,-19.692022,-9.846011,-19.692022,-9.152890,-9.846011,-19.692022)
writeBin(out, "problem.glf", size=8)
' | R --slave
gzip -f problem.glf

echo -e '>chr1\nAAAAAAAAAAA' >problem.fa
sleep 1 && echo -e 'chr1\t0\t10' >problem.fai

git clone https://github.com/ANGSD/angsd.git && cd angsd && make && cd ..
angsd/angsd -glf problem.glf.gz -nind 1 -anc problem.fa -fai problem.fai -domaf 1 -domajorminor 1 -dosaf 1 -out problem &>/dev/null
angsd/misc/realSFS print problem.saf.idx 2>/dev/null
rm problem*
```
This gives
```
chr1    1       -8.054383       0.000000        -9.152943
```
which has the max value for the polymorphic (e.g. heterozygous) SFS bin.

**Comment by TeresaPegan** [comment]
Thanks for checking into it. In your example above, you are using a glf file as input....I didn't actually know you could use a glf file as input for -dosaf 1, I thought based on the [ANGSD docs](http://popgen.dk/angsd/index.php/SFS_Estimation) it had to come straight from the bam files with the -bam argument and the -gl argument. Maybe that's functionally the same thing?When I get a chance I will give it a try with a glf file to see if it fixes this behavior on my end.
Thanks!
-Teresa

**Comment by nspope** [comment]
I'm assuming the GL values you posted above were generated via "-bam X -gl 2 -doglf 4" or something? If that's so, then it looks like they should give identical results aside from any rounding error causes by converting from binary to printed text and back to binary again. But definitely good to double check.

You can also dump binary format straight from angsd (I think) with doglf 1 -- no need to write out to text and convert to binary.

**Comment by TeresaPegan** [comment]
Ok, I have managed to hunt down the problem. It actually has to do with slightly different filtering applied to the code I used to create the "erroneous" SAF file vs the "correct" glf files. The locus in question here had reads "AG" in the raw bam. Evidently the "A" read was being filtered out when I created the initial confusing SAF file, resulting in a SAF file that listed "GG" as the most likely genotype. I'm sorry for not realizing this sooner, it feels like an obvious explanation in hindsight!

I think one reason I didn't think about it earlier is that the filter responsible for this discrepancy is not one of the the baseQ or mapQ filters (which I made sure were the same). Instead it is -C, which is an argument I didn't totally understand. The ANGSD [docs](http://www.popgen.dk/angsd/index.php/Input) say:

-C [int] =0
Adjust mapQ for excessive mismatches (as SAMtools), supply -ref.

Based on samtools' documentation, I guess this removes reads if they show too many mismatches with the reference genome depending on the number you supply. When making the SAF, I was using -C 50 based on advice from a workshop I took.

This is a bit of a eureka moment for me because there have been several occasions in which ANGSD gave me unexpected results that depended a lot on which reference genome I used, with particularly unexpected behavior when the reference genome was from a very distantly related organism (where -C 50 is probably removing huge numbers of reads). I bet this filter was always the cause!
